characters/skills.py

```diff
@@ -1,22 +1,112 @@
-    myCursor.execute("SELECT skillname, points, pk_id FROM skills WHERE fk_owner_id = %s;" % char_pk_id_int)
+    myCursor.execute("SELECT skillname, points, pk_id, fk_owner_id, created_at FROM skills WHERE fk_owner_id = %s;" % char_pk_id_int)
     lines = myCursor.fetchall()
     characters_skills = []
     for line in lines:
         new_skill = parse_line_to_skill(line)
         characters_skills.append(new_skill)
     return characters_skills
 
+""" returns a map from a list such as that expected by the get_characters_skills function 
+    line: the list of skill variables in the order skillname, points, pk_id, fk_owner_id
+        such as that output from a psql query using psychopg2
+"""
 def parse_line_to_skill(line):
     skill = {}
     skill['skill'] = line[0]
     skill['name'] = line[0]
     skill['points'] = line[1]
     skill['pk_id'] = line[2]
-    return skill
\ No newline at end of file
+    skill['fk_owner_id'] = line[3]
+    skill['created_at'] = line[4]
+    return skill
+
+""" sets an existing skill to have the new values """
+def update_skill(new_skill_map):
+    #sanitize!
+    saniname = security.sql_escape(new_skill_map['name'])
+    try:
+        sanipoints = int(new_skill_map['points'])
+    except:
+        return None
+    try:
+        sani_pk_id = int(new_skill_map['pk_id'])
+    except:
+        return None
+    #update
+    connection = characters_common.db_connection()
+    myCursor = connection.cursor()
+    skill_tuple = (saniname, sanipoints, sani_pk_id)
+    myCursor.execute("UPDATE skills SET (skillname, points) = (E'%s', %s) WHERE pk_id = %s;" % (skill_tuple))
+    myCursor.close()
+    connection.commit()
+    return new_skill_map
+
+""" Since Skills hold so little data, and nothing depends on them, we eschew lazy delete with them.
+    pk_id - the primary key of the skill that you wish to delete from the database. """
+def delete_skill(pk_id):
+    try:
+        sani_pk_id = int(pk_id)
+    except:
+        return None
+    connection = characters_common.db_connection()
+    myCursor = connection.cursor()
+    myCursor.execute("UPDATE skills SET deleted_at = now() WHERE pk_id = %s;" % sani_pk_id)
+    return sani_pk_id
+
+""" Called asynchronously to create a new skill for later editing. Returns None if there's an error, otherwise
+    returns the PK_ID of the newly created skill.
+    Expects character_pk_id to be the primary key of the character to which the new skill will be assigned. """
+def get_newest_skill(character_pk_id):
+    try:
+        sani_pk_id = int(character_pk_id)
+    except:
+        return None
+    connection = characters_common.db_connection()
+    myCursor = connection.cursor()
+    myCursor.execute("SELECT max(pk_id) FROM skills;")
+    new_pk_id = int(myCursor.fetchone()[0])
+    new_pk_id = new_pk_id + 1
+    myCursor.execute("INSERT INTO skills (pk_id, fk_owner_id, skillname, points, created_at) VALUES (%s, %s, 'new skill', 0, now());" % (new_pk_id, sani_pk_id))
```

Comment:
does this SQL need the private column in order for us to support that functionality, or are private skills handled elsewhere?

Reply:
at the moment, characters and their items have no sense of private/public. What would that even mean?

Reply:
People experimenting with character ideas they aren't comfortable sharing yet, or DMs' NPCs perhaps?

Reply:
There's no way of looking at anyone else's characters.

Reply:
X|

Reply:
Will approve merge once I get to a pc